### DISCUSSION TITLE
simulate "cls - clear screen" on TDialog - how to can I do this ?

### DISCUSSION THREAD

**Discussion by paule32** [magiblot/tvision#129] [General]
Hello,
I have a TDialog, and I would like insert Text in it.
Currently I use the **draw** member to "render" the Text - but I fail.
So: how can I use DrawBuffer, to fill a field of 80x25, and draw this buffer to the TDialog ?

**Comment by magiblot** [comment]
Hi Jens!

`TDialog` is a `TGroup`, so its purpose is to contain other views. Therefore, if you want to draw your own things you should instead create a new class inheriting from `TView`, insert it into your `TDialog`, and then implement the `draw` member method of this new class.

Try it this way and let me know if you succeed or not.

**Reply by paule32** [reply]
when I use the Code below, I get the attached screen.
```
TOutputWindowChild::TOutputWindowChild(
    TOutputWindow * parent,
    const TRect& bounds):
    TView(bounds),
    owner(parent) {
    
    options |= ofSelectable;
}
void TOutputWindowChild::draw()
{
    int y = 0, ymax = size.y;
    for (y; y < ymax; ++y) {
        writeChar(0, y, 'o', getColor(14), size.x);  // <-- here, I expect 0xe - yellow on black
    }
}
```
![grafik](https://github.com/magiblot/tvision/assets/3682603/af06b3c2-a48d-4633-9474-35dd303ca4da)

**Reply by paule32** [reply]
The pros: I solve my problem with the Code below - I used **mapColor** to make it work.
The cons: I don't know why:
- I have make it twice: use/override the **mapColor**,
- I have to use **getColor**(14);

When I use getColor without mapColor, I get not the final result, I expected (writeChar)

```
TColorAttr
TOutputWindowChild::mapColorOutsideClass(uchar index)
{
    switch (index)
    {
        case  1: return 0x01; break;
        case  2: return 0x02; break;
        case 14: return 0x0e; break;
        case 15: return 0x0f; break;
        
        default:
            return TView::mapColor(index);
    }
}

void TOutputWindowChild::drawContent()
{
    int y = 0, ymax = size.y;
    for (y; y < ymax; ++y) {
        writeChar(0, y, 'o', getColor(14), size.x);
    }
}
```

**Reply by paule32** [reply]
this is now, what I expect:
![grafik](https://github.com/magiblot/tvision/assets/3682603/99dcf464-77d3-46fb-97a6-632aaf49c9c9)

**Comment by magiblot** [comment]
I'll explain what happened:

The `mapColor` and `getColor` methods are related to the palette system. They both get a palette index and return a color attribute (`TColorAttr`: the foreground color, background color and style of a cell). The difference is that `getColor` allows you to retrieve two color attributes at once; that's why it returns an attribute pair (`TAttrPair`).

Your view `TOutputWindowChild` is being inserted onto a `TDialog`. Therefore, when it comes to the palette system, the list of color indices that you can use is that of `TDialog`:

https://github.com/magiblot/tvision/blob/f369a73d835c010fc96f3fae505c73766157d13f/include/tvision/dialogs.h#L39-L75

However, in your example you are using the `TView::writeChar` method, which takes a color index and not a color attribute. In the source code of `TView::writeChar` you can see that it already calls `mapColor`. Therefore, it is wrong to call this method with the value returned by `getColor`. 

This is an example of proper usage of `TView::writeChar`, using the palette index `1`:

https://github.com/magiblot/tvision/blob/f369a73d835c010fc96f3fae505c73766157d13f/source/tvision/textview.cpp#L181

But none of this is actually important for what you are trying to accomplish. You already decided that the color attribute you want is `0x0e` (yellow on black). You don't need the palette system at all. Therefore, you should use color attributes directly instead of palette indices.

In order to do this you should use `TDrawBuffer`, the methods of which take color attributes directly. For example:

```c++
TDrawBuffer b;
for (int y = 0; y < size.y; ++y) {
    b.moveChar(0, 'o', 0x0e, size.x);
    writeLine(0, y, size.x, 1, b);
}
```

This way, you don't have to override nor use `mapColor` and you can forget about palette indices.
